<!-- 
MIT License (modified)

Copyright (c) 2020 The Trustees of the University of Pennsylvania
Authors:
Vasileios Vasilopoulos <vvasilo@seas.upenn.edu>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this **file** (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE. 
-->

<launch>

  <arg name="gui" default="false" />
  <arg name="rviz_gui" default="true" />
  <arg name="paused" default="false" />
  <arg name="software" default="true" />
  <arg name="controller" default="inverse_dynamics" />
  <arg name="live_plot" default="false" />
  <arg name="world" default="flat" />
  <arg name="rotors" default="false" />
  <arg name="tail" default="false" />
  <arg name="tail_type" default="0" />
  <arg name="dira" default="true" />
  <arg name="tail_num" default="0" /> <!-- Tail num should be 1 or 2 if tail:=true -->
  <arg name="x_init" default="6.0" />

  <param name="tail_controller/tail_type" value="$(arg tail_type)" />
  <param name="tail_controller/tail_num" value="$(arg tail_num)" />

  <include file="$(find quad_utils)/launch/mapping.launch">
    <arg name="input_type" value="mesh" />
    <arg name="world" value="$(arg world)" />
  </include>

  <group if="$(arg software)">
    <include file="$(find quad_utils)/launch/visualization.launch">
      <arg name="live_plot" value="$(arg live_plot)" />
      <arg name="rviz_gui" value="$(arg rviz_gui)" />
    </include>
    <include file="$(find quad_utils)/launch/robot_driver.launch">
      <arg name="controller" value="$(arg controller)" />
      <arg name="mocap" value="false" />
      <arg name="is_hw" value="false" />
    </include>
    <!-- <include file="$(find quad_utils)/launch/planning.launch">
      <arg name="global_planner" value="$(arg global_planner)"/>
    </include>

    <group if="$(eval global_planner == 'twist')">
      <node name="twist_key_controller"
        pkg="quad_utils"
        type="twist_key_controller.py"
        respawn="false"
        output="screen">
      </node>
    </group> -->
  </group>

  <!-- Launch the Quad world with specific physics parameters -->
  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="paused" value="$(arg paused)" />
    <arg name="world_name" value="$(find gazebo_scripts)/worlds/$(arg world)/$(arg world).world" />
    <arg name="gui" value="$(arg gui)" />
    <arg name="physics" value="ode" />
    <arg name="verbose" value="$(arg gui)" />
    <arg name="extra_gazebo_args" value="--seed 0" />
  </include>

  <!-- Load the parameters and the sdf description -->
  <include file="$(find quad_utils)/launch/load_params.launch" />

  <group if="$(arg rotors)">
    <!-- use more accurate but slower rotor model -->
    <param name="robot_description_sdf" textfile="$(find spirit_description)/sdf_mesh/spirit_rotors.sdf" />
  </group>
  <group unless="$(arg rotors)">
    <group if="$(arg tail)">
      <group if="$(arg dira)">
        <group if="$(eval arg('tail_num') == 1)">
          <param name="robot_description" textfile="$(find spirit_description)/urdf/spirit_tail_1.urdf" /> <!-- Original: spirit_tail.urdf -->
          <!-- <param name="robot_description" textfile="$(find spirit_description)/urdf/spirit_meshes_tail.urdf" /> -->
          <param name="robot_description_sdf" textfile="$(find spirit_description)/sdf_mesh/spirit_tail_1.sdf" /> <!-- Original: spirit_tail.sdf -->
          <rosparam file="$(find gazebo_scripts)/config/spirit_control_tail_dira.yaml" command="load" /> <!-- Original: spirit_control_tail.yaml -->
        </group>
        <group if="$(eval arg('tail_num') == 2)">
          <param name="robot_description" textfile="$(find spirit_description)/urdf/spirit_tail_2.urdf" />
          <param name="robot_description_sdf" textfile="$(find spirit_description)/sdf_mesh/spirit_tail_2.sdf" />
          <rosparam file="$(find gazebo_scripts)/config/spirit_control_tail_dira_2.yaml" command="load" />
        </group>
      </group>
      <group unless="$(arg dira)">
        <param name="robot_description" textfile="$(find spirit_description)/urdf/spirit_tail.urdf" /> <!-- Original: spirit_tail.urdf -->
        <!-- <param name="robot_description" textfile="$(find spirit_description)/urdf/spirit_meshes_tail.urdf" /> -->
        <param name="robot_description_sdf" textfile="$(find spirit_description)/sdf_mesh/spirit_tail.sdf" /> <!-- Original: spirit_tail.sdf -->
        <rosparam file="$(find gazebo_scripts)/config/spirit_control_tail.yaml" command="load" /> <!-- Original: spirit_control_tail.yaml -->
      </group>
    </group>
    <group unless="$(arg tail)">
      <param name="robot_description" textfile="$(find spirit_description)/urdf/spirit.urdf" /> <!-- Original spirit.urdf --> <!-- URDF for rviz -->
      <param name="robot_description_sdf" textfile="$(find spirit_description)/sdf_mesh/spirit.sdf" /> <!-- Original: spirit.sdf --> <!-- SDF for Gazebo -->
      <rosparam file="$(find gazebo_scripts)/config/spirit_control.yaml" command="load" />
    </group>
  </group>

  <!-- Spawn SDF model -->
  <node name="spawn_sdf_model" pkg="gazebo_ros" type="spawn_model" args="-sdf -param robot_description_sdf -model spirit -x $(arg x_init) -y 0.0 -z 1.25 -Y 1.57" output="screen" />

  <node name="controller_spawner"
    pkg="controller_manager"
    type="spawner"
    respawn="false"
    output="screen"
    ns="/spirit"
    args="joint_controller joint_state_controller"></node>

  <!-- Load the contact state publisher -->
  <node name="contact_state_publisher" pkg="gazebo_scripts" type="contact_state_publisher_node" output="screen" />

</launch>